kep/61-real-load-aware-scheduling/README.md

```diff
@@ -0,0 +1,608 @@
+For the 3rd Goal, since we utilise PodTopologySpread plugin scoring and to prevent double scoring, it is recommended to disable "PodTopologySpread" Score plugin only. The motivation behind using this plugin is explained under "Spreading with BestFitBinPack".
+
+
+### Risks and Mitigations
+
+If utilization metrics are not available for a long time, we will fall back to the best fit bin pack based on allocations. There is no user action needed for this.
+
+
+## Design Details
+
+Our design consists of the following components as outlined in the diagram and described below. We propose two plugins namely "BestFitBinPack" and "SafeBalancing". Both of them use metrics from load watcher for scoring nodes with different algorithms.
+
+<img src="images/design.png" alt="design" width="720" height="628"/>
+
+
+
+### Metrics Provider
+
+This service provides metrics backed by a time-series database — for example, Prometheus, InfluxDB, [Kubernetes Metrics Server](https://github.com/kubernetes-sigs/metrics-server) etc.
+
+
+### Load Watcher/Load Analyser
+
+The load watcher and analyzer both run in a single process. The watcher is responsible for retrieving cluster-wide resource usage metrics like CPU, memory, network, and IO stats over windows of a specified duration from metrics providers above. It stores these in its local cache and persists several aggregations in the host local DB for fault tolerance. The analyzer is responsible for the detection of bad metrics and any remediation. Bad metrics could be due to missing metrics or metrics with considerable errors, making them anomalies. It can also be extended in the future to use ML models for analyzing metrics.
+
+Load watcher would cache metrics in the last 15-minute, 10-minute, and 5-minute windows, which can be queried via REST API exposed. A generic interface will be provided that can support fetching metrics from any metrics provider. Example is provided in Appendix.
+
+
+### DB
+
+This is a localhost database stored as a file. This DB aims to populate the load watcher's cache quickly if it is lost due to a crash and perform scheduling decisions without any impact on latency.
+The file will be stored in host file system, so it will be persisted across pod failures on the same node. For HA setups, each load watcher process will maintain its corresponding file locally.
+
+
+### Scheduler Plugins
+
+This uses the scheduler framework of K8s to incorporate our customized real load aware scheduler plugins without modifying the core scheduler code. The plugins we proposed mainly include the following two.
+
+- BestFitBinPack Plugin: It is a node sorting plugin that sorts nodes by their actual resource utilization in a way that all utilized nodes have around x% of utilization. We use x% = 50% below to describe the plugin.
```

Comment:
I would suggest a name that reflect the "live" nature of these plugins (as opposed to static requests)

Reply:
Something like "LoadAware" prefix added? -> LoadAwareBestFitBinPack, LoadAwareSafeBalancing ?

Reply:
Can we say "RealtimeLoadAware"? A bit verbose though...

I guess "BestFix" doesn't add additional context, can we name it "<prefix to indicate it's based on live data>" + "BinPack"?

Reply:
@Huang-Wei , for safe balancing plugin, we can just call it SafeBalancing as risks can only be balanced with usage variations under consideration.

For BFBP,  it really does more than bin packing. It will do spreading when all nodes' usage exceeding x%. I will leave it to @zorro786.

Besides, we have an agreed name for the whole framework, called "Trimaran". Given that Kubernetes relates to boating, `Trimaran` has three hulls spaced out apart from one another, so it can reduce wave impact, increase stability, and maintain its beam moving forward.

> Can we say "RealtimeLoadAware"? A bit verbose though...
> 
> I guess "BestFix" doesn't add additional context, can we name it "<prefix to indicate it's based on live data>" + "BinPack"?

Reply:
@Huang-Wei "Load Aware" by default refers to real load. We say load balancer and not real load balancer for example. How about "LoadAwareBinPack"? -> As @wangchen615 mentioned it does more than "Best fit" bin pack, it moves to "Least fit" variant of bin pack if usage exceeds x%. Problem with this name is if we want to add "Least Fit" variant plugin later the name will be ambiguous and needs to be changed. So right now can't think of a better name than "LoadAwareBestFitBinPack".....

Reply:
"MetricAwareBestFitBinPack" is an alternative but I prefer "Load"

Reply:
> We say load balancer and not real load balancer for example.

That's fair :) LoadAwareBestFitBinPack sounds good to me. @alculquicondor any thoughts?

Reply:
`LoadAwareBinPack`? The "BestFit" seems kind of implied imo

Reply:
I am fine with `LoadAwareBestFitBinPack` or `LoadAwareBinPack` both. You guys decide and let me know  :sweat_smile: